### PR TITLE
Logging: Allow to hardcode logging factory

### LINKED ISSUE #11423: Logging Provider Selection
Hi,

Elasticsearch tries to [determine which logging "backend"](https://github.com/elastic/elasticsearch/blob/master/src/main/java/org/elasticsearch/common/logging/ESLoggerFactory.java#L33) is available on the classpath.

In my scenario I use the [slf4j-log4j-bridge](https://logging.apache.org/log4j/log4j-2.2/log4j-1.2-api/index.html). The comment at [L36](https://github.com/elastic/elasticsearch/blob/master/src/main/java/org/elasticsearch/common/logging/ESLoggerFactory.java#L36) indicates that in this scenario a NoSuchMethod should be thrown and the Slf4jESLoggerFactory should be used (if slf4j is present).

But the slf4j-log4j-bridge accepts the setLevel(org.apache.log4j.Level). See [org.apache.log4j.Logger](https://git-wip-us.apache.org/repos/asf?p=logging-log4j2.git;a=blob;f=log4j-1.2-api/src/main/java/org/apache/log4j/Logger.java;h=2e643a17c1de79fb56a49710797f046e162de289;hb=HEAD#l26) and [org.apache.log4j.Category](https://git-wip-us.apache.org/repos/asf?p=logging-log4j2.git;a=blob;f=log4j-1.2-api/src/main/java/org/apache/log4j/Category.java;h=25408e79f7ecb61176333df6cbcd9aced3d448ee;hb=HEAD#l190).

### PR DESCRIPTION
**Pull request by spinscale**
Due to the mechanism to automatically select an existing logging
factory on startup this might clash with packages that bridge to
existing logging mechanisms.

In order to be more explicit this commit allows to set the
es.logger.impl property to explicitely configure a JDK, log4j or slf4j
logging factory.

Closes #11423

### CODE REVIEW

**Comment by spinscale** [comment]
thinking about this some more, we could even stop trying to do things automatically, document this new setting and just go with log4j by default...

**Comment by jpountz** [comment]
> thinking about this some more, we could even stop trying to do things automatically, document this new setting and just go with log4j by default...

+1 !

**Comment by spinscale** [comment]
updated the PR, added documentation for this settings, removed the automagic portion of this code

File: core/src/main/java/org/elasticsearch/common/logging/ESLoggerFactory.java [hunk 1]
```diff
@@ -28,22 +28,24 @@
...
-            } catch (Throwable e1) {
-                // no slf4j
-            }
+        defaultFactory = getConfiguredEsLoggerFactory();
+    }
+
+    static ESLoggerFactory getConfiguredEsLoggerFactory() {
+        String loggingType = System.getProperty(LOGGER_IMPL_PROPERTY_NAME, "log4j");
+        switch (loggingType) {
+            case "slf4j":
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/logging/ESLoggerFactory.java]: Argh! Indented `case`es!

**Comment by nik9000** [comment]
I'm trying to figure out what the bridging problem is.

Say you want to use Logback. You'd do this by adding slf4j and log4j-over-slf4j and jcl-over-slf4j to the classpath and elasticsearch would think its logging to log4j because that is the first class it finds. Is the trouble that it'd then try to configure log4j which would fail completely because log4j isn't there?

**Comment by rmuir** [comment]
I am confused by the logging setup myself. Is this another thing where we have conflicting uses of "client" vs "server"?  I am honestly just asking the question, it seems like we can just pick one library that we use, and that is that?

I would love if logging is simpler, I always give up completely if an issue involves logging, I try to avoid use of logging whenever possible in my code, etc because logging is so overengineered and complex.

**Comment by nik9000** [comment]
> I am confused by the logging setup myself. Is this another thing where we have conflicting uses of "client" vs "server"? I am honestly just asking the question, it seems like we can just pick one library that we use, and that is that?

I think so. In the server case right now we support log4j and that is it.

In the client use case we're going to have to conform to whatever the client application wants to use and log4j isn't really a common choice any more. Maybe the problem vanishes in a puff of smoke if we have a java client library.

**Comment by spinscale** [comment]
Yes, this is rather for the client use-case, when someone wants to use another logging framework, which is used by the application. The problem would indeed vanish with a dedicated client.

@nik9000 the current problem is, that you might accidentally use log4j because your bridging library uses the same class. Instead of magically trying to load classes this approach favors explicit configuration. Hope, that explanation makes sense.

**Comment by nik9000** [comment]
> @nik9000 the current problem is, that you might accidentally use log4j because your bridging library uses the same class. Instead of magically trying to load classes this approach favors explicit configuration. Hope, that explanation makes sense.

It makes sense and I'm all for getting it configured for cleanliness but I wanted to be clear on the problem. Is the problem that people using Elasticsearch as a client and using the log4j bridge are triggering elasticsearch to try and configure log4j?

**Comment by rmuir** [comment]
> Yes, this is rather for the client use-case, when someone wants to use another logging framework, which is used by the application. The problem would indeed vanish with a dedicated client.

Can't we still just use one library? We should put the onus on the client to deal with this. There are logging libraries that intercept the api from whichever one we pick and convert it to their logging library of choice.

That is how this should work.

**Comment by nik9000** [comment]
> Can't we still just use one library? We should put the onus on the client to deal with this. There are logging libraries that intercept the api from whichever one we pick and convert it to their logging library of choice.
> 
> That is how this should work.

I think the trouble is the server trying to configure Log4j. We honestly shouldn't do that when elasticsearch is used as a client. If we didn't then we could indeed just target log4j and drop support for the others. People would just use a bridge if they don't like log4j.

**Comment by rmuir** [comment]
> I think the trouble is the server trying to configure Log4j. We honestly shouldn't do that when elasticsearch is used as a client. If we didn't then we could indeed just target log4j and drop support for the others. People would just use a bridge if they don't like log4j.

So lets fix it? This should only happen from Bootstrap.

**Comment by nik9000** [comment]
@spinscale and @rmuir I think this is fixed in master - logging is only set up by Bootstrap and PluginCli.

**Comment by spinscale** [comment]
closing due to #16703

**Comment by nik9000** [comment]
> closing due to #16703

Cool.

**Pull request closed without merging**